File: exts/mf.ov.lidar_live_synth/mf/ov/lidar_live_synth/nodes/OgnBeamToOusterUDPNode.py
"""
This is the implementation of the OGN node defined in OgnBeamToOusterUDPNode.ogn
"""

from pxr import Gf
from typing import List

import math
import numpy as np
import omni.graph.core as og
import socket
import time
import struct

class OgnBeamToOusterUDPNodeInternalState:
    def __init__(self):
        self.frameId = 0

    def update_frameId(self):
        self.frameId += 1

class OgnBeamToOusterUDPNode:
    """
         Receives point cloud data from Isaac Lidar and send it via the Ouster UDP protocol.
    """
    COLUMNS_PER_PACKET = 16
    TWO_PI = np.pi * 2.0
    DEG_TO_RAD = TWO_PI / 360.0
    OUSTER_NUM_ROT_ANGLES = 90112
    OUSTER_NUM_ROT_ANGLES_OVER_TWO_PI = OUSTER_NUM_ROT_ANGLES / TWO_PI

    @staticmethod
    def internal_state():
        """Returns an object that will contain per-node state information"""
        return OgnBeamToOusterUDPNodeInternalState()
    
    @staticmethod
    def compute(db) -> bool:
        """Compute the outputs from the current input"""

        numRows = db.inputs.numRows
        linearDepthData = db.inputs.linearDepthData
        numCols = db.inputs.numCols
        azimuthStart = db.inputs.azimuthRange[0] + OgnBeamToOusterUDPNode.TWO_PI
        horizontalStepInRads = -1.0 * db.inputs.horizontalResolution * OgnBeamToOusterUDPNode.DEG_TO_RAD
        frame_id = db.internal_state.frameId % 65536

        try:
            measurement_id = 0

            if (numRows != 16 and
                  numRows != 32 and
                  numRows != 64 and
                  numRows != 128):

                db.log_error("Row count must be either 16, 32, 64 or 128, not " + str(numRows))
                return False

            # # Packet structure :
            # # OusterAzimuthBlock [16]
            # #     unsigned long long timeStamp
            # #     unsigned short measurementId
            # #     unsigned short frameId
            # #     unsigned int encoderCount
            # #     OusterChannelDataBlock [N] = 16,32,64 OR 128
            # #         unsigned int rangemm
            # #         unsigned short reflectivity
            # #         unsigned short signal_photons
            # #         unsigned short noise_photons
            # #         unsigned short unused
            # #     unsigned int azimuthDataBlockStatus

            packet_point_dtype = np.dtype([('distance', 'u4'),
                                           ('reflectivity', 'u2'),
                                           ('signal_photons', 'u2'),
                                           ('noise_photons', 'u2'),
                                           ('unused', 'u2')])
            
            packet_column_dtype = np.dtype([('timestamp', 'u8'),
                                            ('measurement_id', 'u2'),
                                            ('frame_id', 'u2'),
                                            ('encoder_count', 'u4'),
                                            ('points', packet_point_dtype, (numRows)),
                                            ('block_status', 'u4')])

            chunks = []  # We will likely have multiple packets to send. Each packet contains at most 16 columns
            chunk = np.zeros(OgnBeamToOusterUDPNode.COLUMNS_PER_PACKET, packet_column_dtype)
            current_chunk_column = 0
            
            linearDepthDataInMillimeters = np.zeros(shape=(len(linearDepthData)), dtype=int)
            for index in range(len(linearDepthData)):
                linearDepthDataInMillimeters[index] = int(linearDepthData[index] * 1000)

            # We need to send data in ascending angle (encoder_count) order
            # Data is in right-to-left order, we need to iterate left-to-right
            # We also need to start at the middle (center) of the data which is encoderCount 0
            colEndIndex = (numCols - 1) // 2
            colStartIndex = colEndIndex + numCols

            for temp_col_index in range(colStartIndex, colEndIndex, -1):

                col_index = temp_col_index % numCols

                # This assumes consistent input data across azimuthRange, horizontalResolution, numCols, numRows and linearDepthData size
                # We are lenient on the checks, it should just fail if something doesnt work out and throw an error
                current_encoderCount = int((azimuthStart + horizontalStepInRads * temp_col_index)
                                            * OgnBeamToOusterUDPNode.OUSTER_NUM_ROT_ANGLES_OVER_TWO_PI)

                # If previous chunk is complete, start new one
                if current_chunk_column == OgnBeamToOusterUDPNode.COLUMNS_PER_PACKET:
                    chunks.append(chunk)
                    chunk = np.zeros(OgnBeamToOusterUDPNode.COLUMNS_PER_PACKET, packet_column_dtype)
                    current_chunk_column = 0
                
                chunk[current_chunk_column]['timestamp'] = time.time_ns()
                chunk[current_chunk_column]['measurement_id'] = measurement_id
                chunk[current_chunk_column]['frame_id'] = frame_id
                chunk[current_chunk_column]['encoder_count'] = current_encoderCount

                measurement_id = (measurement_id + 1) % 65536
                col_index_start = col_index * numRows

                for row_index in range(numRows):

                    chunk[current_chunk_column]['points'][row_index]['distance'] = linearDepthDataInMillimeters[col_index_start + row_index]
                    chunk[current_chunk_column]['points'][row_index]['signal_photons'] = 0xFFFF #0xFFFF means valid

                chunk[current_chunk_column]['block_status'] = 0xFFFFFFFF #0xFFFFFFFF means valid

                current_chunk_column += 1

            if current_chunk_column != 0:
                for extra_column_index in range(current_chunk_column, OgnBeamToOusterUDPNode.COLUMNS_PER_PACKET):
                    chunk[extra_column_index]['timestamp'] = time.time_ns()
                    chunk[extra_column_index]['measurement_id'] = measurement_id
                    chunk[extra_column_index]['frame_id'] = frame_id
                    chunk[extra_column_index]['encoder_count'] = OgnBeamToOusterUDPNode.OUSTER_NUM_ROT_ANGLES
                
                chunks.append(chunk)

            if db.inputs.broadcast:
                sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
                sock.setsockopt(socket.IPPROTO_IP, socket.IP_MULTICAST_TTL, 32)
                for chunk in chunks:
                    sock.sendto(chunk, (db.inputs.ip_address, db.inputs.port))
            else:
                sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
                for chunk in chunks:
                    sock.sendto(chunk, (db.inputs.ip_address, db.inputs.port))
            
        except Exception as error:
            # If anything causes your compute to fail report the error and return False
            db.log_error(str(error))
            return False
        
        db.internal_state.update_frameId()

        # Always enable the output execution
        db.outputs.execOut = og.ExecutionAttributeState.ENABLED
        # Even if inputs were edge cases like empty arrays, correct outputs mean success

        return True

File: source/extensions/mf.ov.lidar_live_synth/plugins/nodes/OgnBeamToOusterUDPNode.cpp
#include <OgnBeamToOusterUDPNodeDatabase.h>

#include <chrono>

#define WIN32_LEAN_AND_MEAN
#define _WINSOCK_DEPRECATED_NO_WARNINGS
#ifdef _WIN32
#include <Winsock2.h>
#else
#include <arpa/inet.h>
#include <netdb.h>
#include <netinet/in.h>
#include <sys/select.h>
#include <sys/socket.h>
#include <sys/types.h>
#define SOCKET int
#define INVALID_SOCKET  (SOCKET)(~0)
#define SOCKET_ERROR            (-1)
#define closesocket close
#define SOCKADDR sockaddr
#endif


namespace mf {
namespace ov {
namespace lidar_live_synth {

static const int kColumnsPerPacket = 16;
static const float kPi = 3.14159265359f;
static const float kTwoPi = kPi * 2.0f;
static const float kDegToRad = kTwoPi / 360.0f;
static const int kOusterNumRotAngles = 90112;
static const float kOusterNumRotAnglesOverTwoPi = kOusterNumRotAngles / kTwoPi;

class OgnBeamToOusterUDPNode
{
    int m_frameId{ 0 };

#pragma pack(push,4) // Force packing in 4-byte packs (Words)

    struct OusterChannelDataBlock
    {
        unsigned int rangemm;
        unsigned short reflectivity;
        unsigned short signal_photons;
        unsigned short noise_photons;
        unsigned short unused;

        OusterChannelDataBlock()
            : rangemm(0)
            , reflectivity(0)
            , signal_photons(0)
            , noise_photons(0)
            , unused(0)
        {}
    };

    template <int NUMROWS>
    struct OusterAzimuthBlock
    {
        unsigned long long timeStamp;					  // Word 0,1
        unsigned short measurementId;					  // Word 2[0:15]
        unsigned short frameId;							  // Word 2[16:31]
        unsigned int encoderCount;						  // Word 3
        OusterChannelDataBlock channelDataBlock[NUMROWS]; // Word [4:195] in groups of 3
        unsigned int azimuthDataBlockStatus;			  // word 196

        OusterAzimuthBlock()
            : timeStamp(0)
            , measurementId(0)
            , frameId(0)
            , encoderCount(0)
            , channelDataBlock{}
            , azimuthDataBlockStatus(0)
        {}
    };

    template <int NUMROWS>
    struct OusterDataPacket
    {
        OusterAzimuthBlock<NUMROWS> block[16]; // Each packet consists of 16 azimuth blocks

        OusterDataPacket()
            :block{}
        {}
    };

#pragma pack(pop)


    class OgnBeamToOusterUDPNodeSocket
    {
    public:
        OgnBeamToOusterUDPNodeSocket()
            : SendSocket(INVALID_SOCKET)
            , isBroadcastSocket(false)
        {}

        virtual ~OgnBeamToOusterUDPNodeSocket()
        {
            if (SendSocket != INVALID_SOCKET)
            {
                closesocket(SendSocket);
            }
        }

        bool prepare(OgnBeamToOusterUDPNodeDatabase& db)
        {
            if (isBroadcastSocket != db.inputs.broadcast())
            {
                closesocket(SendSocket);
                SendSocket = INVALID_SOCKET;
            }

            if (SendSocket == INVALID_SOCKET)
            {
                SendSocket = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);

                if (SendSocket == INVALID_SOCKET)
                {
                    db.logError("Error in OgnBeamToOusterUDPNode opening socket : %d", SendSocket);
                    return false;
                }

                if (db.inputs.broadcast())
                {
                    char broadcast = 1;
                    int iResult = setsockopt(SendSocket, SOL_SOCKET, SO_BROADCAST, &broadcast, sizeof(broadcast));
                    if (!iResult)
                    {
                        closesocket(SendSocket);
                        SendSocket = INVALID_SOCKET;

                        db.logError("Error in OgnBeamToOusterUDPNode setting socket options : %d", iResult);
                        return false;
                    }
                }

                isBroadcastSocket = db.inputs.broadcast();
            }

            RecvAddr.sin_family = AF_INET;
            RecvAddr.sin_port = htons(db.inputs.port());
            std::string ipAddress = db.inputs.ip_address();
            RecvAddr.sin_addr.s_addr = inet_addr(ipAddress.data());

            return true;
        }

        template <int NUMROWS>
        bool send(const OusterDataPacket<NUMROWS>& packet, OgnBeamToOusterUDPNodeDatabase& db)
        {
            int iResult = sendto(SendSocket, reinterpret_cast<const char*>(&packet), sizeof(packet), 0, (SOCKADDR*)&RecvAddr, sizeof(RecvAddr));
            if (iResult == SOCKET_ERROR)
            {
                db.logError("Error in OgnBeamToOusterUDPNode sending data on socket : %d", iResult);

                return false;
            }

            return true;
        }
    private:
        SOCKET SendSocket;
        sockaddr_in RecvAddr;
        bool isBroadcastSocket;
    };

    OgnBeamToOusterUDPNodeSocket m_ognBeamToOusterUDPNodeSocket;

    template<int NUMROWS>
    static bool computeForSize(OgnBeamToOusterUDPNodeDatabase& db)
    {
        auto& state = db.internalState<OgnBeamToOusterUDPNode>();

        const auto& linearDepthData = db.inputs.linearDepthData();
        const int& numCols = db.inputs.numCols();
        const float& azimuthStart = db.inputs.azimuthRange()[0] + kTwoPi + kTwoPi;
        const float& horizontalStepInRads = -1.0f * db.inputs.horizontalResolution() * kDegToRad;
        const int& frameId = state.m_frameId % 65536;

        try
        {
            if (!state.m_ognBeamToOusterUDPNodeSocket.prepare(db))
            {
                return false;
            }

            int measurementId = 0;

            OusterDataPacket<NUMROWS> packet;
            int currentChunkColumn = 0;

            // We need to send data in ascending angle (encoder_count) order
            // Data is in right-to-left order, we need to iterate left-to-right
            // We also need to start at the middle (center) of the data which is encoderCount 0
            int colEndIndex = (numCols - 1) / 2;
            int colStartIndex = colEndIndex + numCols;

            for (int tempColIndex = colStartIndex; tempColIndex > colEndIndex; tempColIndex--)
            {
                int colIndex = tempColIndex % numCols;

                // This assumes consistent input data across azimuthRange, horizontalResolution, numCols, numRows and linearDepthData size
                int currentEncoderCount = int((azimuthStart + horizontalStepInRads * tempColIndex) * kOusterNumRotAnglesOverTwoPi);
                if (currentEncoderCount < 0 || currentEncoderCount >= kOusterNumRotAngles)
                {
                    db.logError("currentEncoderCount must be between 0 and %d, not %d", kOusterNumRotAngles, currentEncoderCount);
                    return false;
                }

                // If previous chunk is complete, start new one
                if (currentChunkColumn == kColumnsPerPacket)
                {
                    state.m_ognBeamToOusterUDPNodeSocket.send<NUMROWS>(packet, db);
                    packet = OusterDataPacket<NUMROWS>();

                    currentChunkColumn = 0;
                }

                packet.block[currentChunkColumn].timeStamp =
                    std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::system_clock::now().time_since_epoch()).count();
                packet.block[currentChunkColumn].measurementId = measurementId;
                packet.block[currentChunkColumn].frameId = frameId;
                packet.block[currentChunkColumn].encoderCount = currentEncoderCount;

                measurementId = (measurementId + 1) % 65536;
                int colIndexStart = colIndex * NUMROWS;

                for (int rowIndex = 0; rowIndex < NUMROWS; rowIndex++)
                {
                    packet.block[currentChunkColumn].channelDataBlock[rowIndex].rangemm = (int)(linearDepthData[colIndexStart + rowIndex] * 1000.0f);
                    packet.block[currentChunkColumn].channelDataBlock[rowIndex].signal_photons = 0xFFFF; //0xFFFF means valid
                }

                packet.block[currentChunkColumn].azimuthDataBlockStatus = 0xFFFFFFFF; //0xFFFFFFFF means valid

                currentChunkColumn++;
            }

            if (currentChunkColumn != 0)
            {
                for (int extraColumnIndex = currentChunkColumn; extraColumnIndex < kColumnsPerPacket; extraColumnIndex++)
                {
                    packet.block[extraColumnIndex].timeStamp =
                        std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::system_clock::now().time_since_epoch()).count();
                    packet.block[extraColumnIndex].measurementId = measurementId;
                    packet.block[extraColumnIndex].frameId = frameId;
                    packet.block[extraColumnIndex].encoderCount = kOusterNumRotAngles;
                }

                state.m_ognBeamToOusterUDPNodeSocket.send<NUMROWS>(packet, db);
            }

        }
        catch (...)
        {
            db.logError("Error in OgnBeamToOusterUDPNode::compute");
            return false;

        }

        state.m_frameId++;

        // Always enable the output execution
        db.outputs.execOut() = omni::graph::core::ExecutionAttributeState::kExecutionAttributeStateEnabled;

        // Even if inputs were edge cases like empty arrays, correct outputs mean success
        return true;
    }

public:
    static bool compute(OgnBeamToOusterUDPNodeDatabase& db)
    {
        // TODO: why is state declared here
        // auto& state = db.internalState<OgnBeamToOusterUDPNode>();
        const int& numRows = db.inputs.numRows();

        switch (numRows)
        {
        case 16:
            return computeForSize<16>(db);
            break;
        case 32:
            return computeForSize<32>(db);
            break;
        case 64:
            return computeForSize<64>(db);
            break;
        case 128:
            return computeForSize<128>(db);
            break;
        }

        db.logError("Row count must be either 16, 32, 64 or 128, not %d", numRows);
        return false;
    }
};

// This macro provides the information necessary to OmniGraph that lets it automatically register and deregister
// your node type definition.
REGISTER_OGN_NODE()

}
}
}
